src/main/java/com/maxdemarzi/shortest/Dijkstra.java

```diff
@@ -0,0 +1,179 @@
+package com.maxdemarzi.shortest;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map.Entry;
+import java.util.PriorityQueue;
+import java.util.Map;
+
+import org.neo4j.cursor.Cursor;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.kernel.api.ReadOperations;
+import org.neo4j.kernel.api.cursor.NodeItem;
+import org.neo4j.kernel.api.cursor.RelationshipItem;
+
+import net.openhft.koloboke.collect.map.IntIntMap;
+import net.openhft.koloboke.collect.map.IntIntMapFactory;
+import net.openhft.koloboke.collect.map.hash.HashIntIntMap;
+import net.openhft.koloboke.collect.map.hash.HashIntIntMaps;
+import net.openhft.koloboke.collect.map.hash.HashLongIntMap;
+import net.openhft.koloboke.collect.map.hash.HashLongLongMap;
+import net.openhft.koloboke.collect.map.hash.HashLongLongMaps;
+import net.openhft.koloboke.collect.map.hash.HashLongObjMap;
+import net.openhft.koloboke.collect.map.hash.HashLongObjMaps;
+import net.openhft.koloboke.collect.set.hash.HashLongSet;
+import net.openhft.koloboke.collect.set.hash.HashLongSets;
+import net.openhft.koloboke.function.Consumer;
+import net.openhft.koloboke.function.IntIntConsumer;
+
+public final class Dijkstra {
```

Comment:
one concern with this algorithm is that it goes 4 degrees from the start nodes, and 0 degrees from the edge nodes. the existing BFS goes 3 degrees from the start nodes, and 1 degree from the edge nodes, to avoid that explosion at 4 degrees. i think making that change was the thing that stopped our extension from killing neo4j. maybe we have enough additional spare capacity now to handle going 4 degrees, but i'm worried about it, and i think we should at least be thinking about how this algorithm might work if we want to do part of the search in reverse.